File: cpp/BST.cpp
struct Node {
  int data;
  bool color;
  Node *left, *right, *parent;

  // 构造器
  Node(int data) {
    this->data = data;
    left = right = parent = NULL;
    this->color = RED;
  }
};

/*标准的二叉排序树的插入操作*/
Node *BSTInsert(Node *root, Node *x) {
  /* 如果树为空，则插入的结点作为根结点返回 */
  if (root == NULL)
    return x;

  /* x的值小于根结点的值，则向左子树走 */
  if (x->data < root->data) {
    root->left = BSTInsert(root->left, x);
    root->left->parent = root;
  } else if (x->data > root->data) {
    root->right = BSTInsert(root->right, x);
    root->right->parent = root;
  }

  /* 返回插入结点 x 的树根*/
  return root;
}

void RBTree::rotateLeft(Node *&root, Node *&g) {
  Node *p = g->right;

  g->right = p->left;  //g->right = T3

  if (g->right != NULL)    //T3 != NULL
    g->right->parent = g;  //T3->parent = g

  p->parent = g->parent;

  if (g->parent == NULL) {
    root = p;
  } else if (g == g->parent->left) {
    g->parent->left = p;
  } else {
    g->parent->right = p;
  }

  p->left = g;
  g->parent = p;
}

void RBTree::rotateRight(Node *&root, Node *&g) {
  Node *p = g->left;

  g->left = p->right;

  if (g->left != NULL)
    g->left->parent = g;

  p->parent = g->parent;

  if (g->parent == NULL)
    root = p;

  else if (g == g->parent->left)
    g->parent->left = p;

  else
    g->parent->right = p;

  p->right = g;
  g->parent = p;
}
// 处理由于红黑树结点的插入操作所造成的平衡问题
void RBTree::fixViolation(Node *&root, Node *&pt) {
  Node *parent_pt = NULL;
  Node *grand_parent_pt = NULL;

  while ((pt != root) && (pt->color != BLACK) &&
         (pt->parent->color == RED)) {
    parent_pt = pt->parent;
    grand_parent_pt = pt->parent->parent;

    /* 情况一:
   pt的父结点parent_pt 是其爷爷结点grand_parent_pt的左孩子 */
    if (parent_pt == grand_parent_pt->left) {
      Node *uncle_pt = grand_parent_pt->right;

      /* Case : a) pt 的叔叔结点 uncle_pt 为红色结点
      只需要调整结点颜色，不需要旋转*/
      if (uncle_pt != NULL && uncle_pt->color == RED) {
        grand_parent_pt->color = RED;
        parent_pt->color = BLACK;
        uncle_pt->color = BLACK;
        pt = grand_parent_pt;
      } else  // Case :b) x 的叔叔结点为黑色
      {
        /* Case : 2
    pt 是其父结点的右孩子
    需要左旋操作 */
        if (pt == parent_pt->right) {
          rotateLeft(root, parent_pt);
          pt = parent_pt;
          parent_pt = pt->parent;
        }

        /* Case : 3
    pt 是其父结点的左孩子
    需要右旋操作 */
        rotateRight(root, grand_parent_pt);
        swap(parent_pt->color, grand_parent_pt->color);
        pt = parent_pt;
      }
    }

    /* 情况二:
  pt的父结点parent_pt 是其爷爷结点grand_parent_pt的左孩子 */
    else {
      Node *uncle_pt = grand_parent_pt->left;

      /* Case : a) pt 的叔叔结点 uncle_pt 为红色结点
      只需要调整结点颜色，不需要旋转*/
      if ((uncle_pt != NULL) && (uncle_pt->color == RED)) {
        grand_parent_pt->color = RED;
        parent_pt->color = BLACK;
        uncle_pt->color = BLACK;
        pt = grand_parent_pt;
      } else  // Case :b) x 的叔叔结点为黑色
      {
        /* Case : 2
    pt 是其父结点的右孩子
    需要左旋操作 */
        if (pt == parent_pt->left) {
          rotateRight(root, parent_pt);
          pt = parent_pt;
          parent_pt = pt->parent;
        }

        /* Case : 3
    pt 是其父结点的左孩子
    需要右旋操作 */
        rotateLeft(root, grand_parent_pt);
        swap(parent_pt->color, grand_parent_pt->color);
        pt = parent_pt;
      }
    }
  }

  //如果是根结点，颜色置为黑色
  root->color = BLACK;
}

// 红黑树插入操作
void RBTree::insert(const int &data) {
  Node *pt = new Node(data);

  // 执行标准的BST插入操作
  root = BSTInsert(root, pt);

  // 进行平衡处理
  fixViolation(root, pt);
}

File: PHP/algorithms/Tree/BST.php
<?php


namespace Algorithms\Tree;

class Node
{

	public $data;
	public $left;
	public $right;
	public $parent;

	public function __construct(int $data = null, Node $parent = null)
	{
		$this->data = $data;
		$this->parent = $parent;
		$this->left = null;
		$this->right = null;
	}

	public function min()
	{
		$node = $this;

		while ($node->left) {
			$node = $node->left;
		}

		return $node;
	}

	public function max()
	{
		$node = $this;

		while ($node->right) {
			$node = $node->right;
		}

		return $node;
	}

	public function successor()
	{

		$node = $this;
		if ($node->right) {
			return $node->right->min();
		} else {
			return null;
		}
	}

	public function predecessor()
	{
		$node = $this;
		if ($node->left) {
			return $node->left->max();
		} else {
			return null;
		}
	}

	public function delete()
	{
		$node = $this;
		if (!$node->left && !$node->right) {
			if ($node->parent->left === $node) {
				$node->parent->left = null;
			} else {
				$node->parent->right = null;
			}
		} elseif ($node->left && $node->right) {
			$successor = $node->successor();
			$node->data = $successor->data;
			$successor->delete();
		} elseif ($node->left) {
			if ($node->parent->left === $node) {
				$node->parent->left = $node->left;
				$node->left->parent = $node->parent->left;
			} else {
				$node->parent->right = $node->left;
				$node->left->parent = $node->parent->right;
			}
			$node->left = null;
		} elseif ($node->right) {

			if ($node->parent->left === $node) {
				$node->parent->left = $node->right;
				$node->right->parent = $node->parent->left;
			} else {
				$node->parent->right = $node->right;
				$node->right->parent = $node->parent->right;
			}
			$node->right = null;
		}
	}

}

class BST
{

	public $root = null;

	public function __construct(int $data)
	{
		$this->root = new Node($data);
	}

	public function isEmpty(): bool
	{
		return $this->root === null;
	}

	public function insert(int $data)
	{
		if ($this->isEmpty()) {
			$node = new Node($data);
			$this->root = $node;
			return $node;
		}
		$node = $this->root;
		while ($node) {
			if ($data > $node->data) {
				if ($node->right) {
					$node = $node->right;
				} else {
					$node->right = new Node($data, $node);
					$node = $node->right;
					break;
				}
			} elseif ($data < $node->data) {
				if ($node->left) {
					$node = $node->left;
				} else {
					$node->left = new Node($data, $node);
					$node = $node->left;
					break;
				}
			} else {
				break;
			}
		}
		return $node;
	}

	public function traverse(Node $node, string $type = "in-order")
	{
		switch ($type) {
			case "in-order":
				$this->inOrder($node);
				break;
			case "pre-order":
				$this->preOrder($node);
				break;
			case "post-order":
				$this->postOrder($node);
				break;
		}
	}

	public function preOrder(Node $node)
	{
		if ($node) {
			echo $node->data." ";
			if ($node->left) {
				$this->traverse($node->left);
			}
			if ($node->right) {
				$this->traverse($node->right);
			}
		}
	}

	public function inOrder(Node $node)
	{
		if ($node) {
			if ($node->left) {
				$this->traverse($node->left);
			}
			echo $node->data." ";
			if ($node->right) {
				$this->traverse($node->right);
			}
		}
	}

	public function postOrder(Node $node)
	{
		if ($node) {
			if ($node->left) {
				$this->traverse($node->left);
			}
			if ($node->right) {
				$this->traverse($node->right);
			}
		}
	}

	public function search(int $data)
	{
		if ($this->isEmpty()) {
			return false;
		}
		$node = $this->root;
		while ($node) {
			if ($data > $node->data) {
				$node = $node->right;
			} elseif ($data < $node->data) {
				$node = $node->left;
			} else {
				break;
			}
		}
		return $node;
	}


	public function remove(int $data)
	{
		$node = $this->search($data);
		if ($node) {
			$node->delete();
		}
	}
}

$tree = new BST(10);
$tree->insert(12);
$tree->insert(6);
$tree->insert(3);
$tree->insert(8);
$tree->insert(15);
$tree->insert(13);
$tree->insert(36);
$tree->traverse($tree->root);
echo PHP_EOL;

echo $tree->search(14) ? "Found"."\n" : "Not Found"."\n";
echo $tree->search(36) ? "Found"."\n" : "Not Found"."\n";
$tree->remove(15);
$tree->traverse($tree->root);
echo PHP_EOL;

$tree->traverse($tree->root, 'pre-order');

echo PHP_EOL;
$tree->traverse($tree->root, 'in-order');

echo PHP_EOL;
$tree->traverse($tree->root, 'post-order');
